embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/session/orchestrator/SessionOrchestratorImpl.kt

```diff
@@ -171,6 +189,66 @@
         }
     }
 
+    /**
+     * This function is responsible for transitioning state from one session to another. This can
+     * be summarised in 3 steps:
+     *
+     * 1. End the current session or background activity.
+     * 2. Clean up any previous session state.
+     * 3. Start the next session or background activity.
+     *
+     * @param description       A key used in logs to identify the transition
+     * @param timestamp         The timestamp of the transition.
+     * @param endProcessState   The process state (foreground/background) upon ending the transition
+     * @param oldSessionAction  The action that ends the old session or background activity (if any).
+     * The initial session object (if any) is passed as a parameter to allow building a full payload.
+     * @param newSessionAction  The action that starts the new session or background activity (if any).
+     * If a new session is created this must return a session object containing the initial state.
+     * @param clearUserInfo     Whether to clear user info when ending the session. Defaults to false
+     */
+    private fun transitionState(
+        description: String,
+        timestamp: Long,
+        endProcessState: State,
+        oldSessionAction: (initial: Session) -> Unit,
+        newSessionAction: () -> Session?,
+        earlyTerminationCondition: () -> Boolean = { false },
+        clearUserInfo: Boolean = false,
+    ) {
+        // supplied business logic says that we can't perform a transition yet.
+        // exit early & retain the current state instead.
+        if (earlyTerminationCondition()) {
+            return
```

Comment:
Is there utility in returning whether this succeeded or not? Might be an interesting case to log an internal error/warning for if this is unexpected

Reply:
I've added some extra logging for this in this specific class. There are already some existing internal errors logged from ProcessStateService if an unbalanced call happens.